openprescribing/frontend/price_per_unit/substitution_sets.py

```diff
@@ -0,0 +1,274 @@
+)
+
+
+# BNF codes which we can't map to generics for various reasons. Issue numbers
+# below refer to https://github.com/ebmdatalab/price-per-dose/issues
+EXCLUSIONS_RE = re.compile(
+    r"""
+    (
+      0302000C0....BE |  # issue #10
+      0302000C0....BF |  # issue #10
+      0302000C0....BH |  # issue #10
+      0302000C0....BG |  # issue #10
+      0904010H0.*     |  # issue #9
+      1311070S0....AA |  # issue #9
+      1311020L0....BS |  # issue #9
+      0301020S0....AA |  # issue #12
+      190700000BBCJA0 |  # issue #12
+      0604011L0BGAAAH |  # issue #12
+      1502010J0....BY |  # issue #12
+      1201010F0AAAAAA |  # issue #12
+      0107010S0AAAGAG |  # issue #12
+      060016000BBAAA0 |  # issue #14
+      190201000AABJBJ |  # issue #14
+      190201000AABKBK |  # issue #14
+      190201000AABLBL |  # issue #14
+      190201000AABMBM |  # issue #14
+      190201000AABNBN |  # issue #14
+      190202000AAADAD    # issue #14
+    )
+    """,
+    re.VERBOSE,
+)
+
+
+# These are BNF code prefixes where all presentations within the sub-paragraph
+# can be considered substitutable. See
+# https://github.com/ebmdatalab/price-per-dose/issues/1.
+GENERIC_SUB_PARAGRAPHS = {
+    "0601060U0": "Urine Testing Reagents",
+    "0601060D0": "Glucose Blood Testing Reagents",
+}
+
+
+@memoize
+def get_substitution_sets():
+    bnf_codes = [row[0] for row in get_db().query("SELECT bnf_code FROM presentation")]
+    substitution_sets = get_substitution_sets_from_bnf_codes(bnf_codes)
+    return {s.id: s for s in substitution_sets}
+
+
+def get_substitution_sets_from_bnf_codes(bnf_codes):
+    """
+    Given a list of BNF codes return a list of SubstitutionSets over those BNF
+    codes
+    """
+    swaps, swap_descriptions = get_formulation_swaps(FORMULATION_SWAPS_FILE)
+    presentation_sets = defaultdict(list)
+    for bnf_code in bnf_codes:
+        generic_code = generic_equivalent_for_bnf_code(bnf_code)
+        if generic_code:
+            generic_code = swaps.get(generic_code, generic_code)
+            presentation_sets[generic_code].append(bnf_code)
+    names = get_names_for_bnf_codes(presentation_sets.keys())
+    return [
+        SubstitutionSet(
+            id=code,
+            name=names.get(code, "unknown"),
+            presentations=sorted(presentations),
+            formulation_swaps=swap_descriptions.get(code),
+        )
+        for code, presentations in presentation_sets.items()
+        # There's no point in a substitution set with only one member
+        if len(presentations) > 1
+    ]
+
+
+def get_names_for_bnf_codes(bnf_codes):
+    names = Presentation.names_for_bnf_codes(list(bnf_codes))
+    # Add in names for our invented BNF codes which represent their entire BNF
+    # sub-paragraph
+    for sub_para, name in GENERIC_SUB_PARAGRAPHS.items():
+        names[sub_para + "AAA0A0"] = name
+    return names
+
+
+def generic_equivalent_for_bnf_code(bnf_code):
+    """
```

Comment:
Possibly some opportunity for reusing this in the new analyse form stuff you're working on @inglesp? If so perhaps belongs in a more-generic module?